pkg/sql/distsqlplan/aggregator_funcs.go

```diff
@@ -16,54 +16,142 @@
 	},
 
 	distsqlrun.AggregatorSpec_BOOL_AND: {
-		LocalStage: distsqlrun.AggregatorSpec_BOOL_AND,
-		FinalStage: distsqlrun.AggregatorSpec_BOOL_AND,
+		LocalStage: []distsqlrun.AggregatorSpec_Func{distsqlrun.AggregatorSpec_BOOL_AND},
+		FinalStage: []distsqlrun.AggregatorSpec_Func{distsqlrun.AggregatorSpec_BOOL_AND},
 	},
 
 	distsqlrun.AggregatorSpec_BOOL_OR: {
-		LocalStage: distsqlrun.AggregatorSpec_BOOL_OR,
-		FinalStage: distsqlrun.AggregatorSpec_BOOL_OR,
+		LocalStage: []distsqlrun.AggregatorSpec_Func{distsqlrun.AggregatorSpec_BOOL_OR},
+		FinalStage: []distsqlrun.AggregatorSpec_Func{distsqlrun.AggregatorSpec_BOOL_OR},
 	},
 
 	distsqlrun.AggregatorSpec_COUNT: {
-		LocalStage: distsqlrun.AggregatorSpec_COUNT,
-		FinalStage: distsqlrun.AggregatorSpec_SUM_INT,
+		LocalStage: []distsqlrun.AggregatorSpec_Func{distsqlrun.AggregatorSpec_COUNT},
+		FinalStage: []distsqlrun.AggregatorSpec_Func{distsqlrun.AggregatorSpec_SUM_INT},
 	},
 
 	distsqlrun.AggregatorSpec_MAX: {
-		LocalStage: distsqlrun.AggregatorSpec_MAX,
-		FinalStage: distsqlrun.AggregatorSpec_MAX,
+		LocalStage: []distsqlrun.AggregatorSpec_Func{distsqlrun.AggregatorSpec_MAX},
+		FinalStage: []distsqlrun.AggregatorSpec_Func{distsqlrun.AggregatorSpec_MAX},
 	},
 
 	distsqlrun.AggregatorSpec_MIN: {
-		LocalStage: distsqlrun.AggregatorSpec_MIN,
-		FinalStage: distsqlrun.AggregatorSpec_MIN,
+		LocalStage: []distsqlrun.AggregatorSpec_Func{distsqlrun.AggregatorSpec_MIN},
+		FinalStage: []distsqlrun.AggregatorSpec_Func{distsqlrun.AggregatorSpec_MIN},
 	},
 
 	distsqlrun.AggregatorSpec_SUM: {
-		LocalStage: distsqlrun.AggregatorSpec_SUM,
-		FinalStage: distsqlrun.AggregatorSpec_SUM,
+		LocalStage: []distsqlrun.AggregatorSpec_Func{distsqlrun.AggregatorSpec_SUM},
+		FinalStage: []distsqlrun.AggregatorSpec_Func{distsqlrun.AggregatorSpec_SUM},
+	},
+
+	// AVG is more tricky than the ones above; we need two intermediate values in
+	// the local and final stages:
+	//  - the local stage accumulates the SUM and the COUNT;
+	//  - the final stage sums these partial results (SUM and SUM_INT);
+	//  - a final rendering then divides the two results.
+	//
+	// At a high level, this is analogous to rewriting AVG(x) as SUM(x)/COUNT(x).
+	distsqlrun.AggregatorSpec_AVG: {
+		LocalStage: []distsqlrun.AggregatorSpec_Func{
+			distsqlrun.AggregatorSpec_SUM,
+			distsqlrun.AggregatorSpec_COUNT,
+		},
+		FinalStage: []distsqlrun.AggregatorSpec_Func{
+			distsqlrun.AggregatorSpec_SUM,
+			distsqlrun.AggregatorSpec_SUM_INT,
+		},
+		FinalRendering: func(h *parser.IndexedVarHelper, varIdxOffset int) (parser.TypedExpr, error) {
+			sum := h.IndexedVar(varIdxOffset)
+			count := h.IndexedVar(varIdxOffset + 1)
+
+			expr := &parser.BinaryExpr{
+				Operator: parser.Div,
+				Left:     sum,
+				Right:    count,
+			}
+
+			if sum.ResolvedType().Equivalent(parser.TypeFloat) {
+				// There is no "FLOAT / INT" operator; cast the denominator to float.
```

Comment:
Might want to mention that there is a "DECIMAL/INT" operator, so we don't need the same handling for that case.